<a href="https://colab.research.google.com/github/GregoryG3/Thesis/blob/main/INITIALIZATION.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# INITIALIZATION

# Importing necessary libraries

In [1]:
!pip install python-docx

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 244.3/244.3 kB 4.5 MB/s eta 0:00:00


In [2]:
import pandas as pd
import numpy as np
from datetime import datetime

from google.colab import drive
from os.path import join

import re
from scipy.stats import describe
# import krippendorff
from statsmodels.stats.inter_rater import fleiss_kappa, aggregate_raters

from matplotlib import pyplot as plt
import seaborn as sns

from docx import Document
from docx.shared import Pt
from docx.enum.text import WD_PARAGRAPH_ALIGNMENT
from docx.oxml.ns import qn
from docx.oxml import OxmlElement

# import random
# import os


# from transformers import (AutoTokenizer, AutoModelForSequenceClassification, BertForSequenceClassification,
#                           BertTokenizer, AdamW, Trainer, TrainingArguments, XLNetTokenizer, XLNetForSequenceClassification,
#                           RobertaTokenizer, RobertaForSequenceClassification, AutoModelForSeq2SeqLM, pipeline)
# import torch
# from torch.utils.data import Dataset, TensorDataset, DataLoader, RandomSampler, SequentialSampler
# from tqdm import tqdm

# from sklearn.metrics import (f1_score, classification_report, confusion_matrix, ConfusionMatrixDisplay,
                            #  accuracy_score, precision_score, recall_score)
# from sklearn.model_selection import train_test_split
# from sklearn.preprocessing import LabelEncoder

# import string
# import nltk
# from nltk.util import ngrams
# from nltk.corpus import stopwords
# from collections import Counter
# from bs4 import BeautifulSoup


# from matplotlib.lines import Line2D
# from wordcloud import WordCloud
# # from PIL import Image

# Load data

In [4]:
drive.mount('/content/drive')
PROJECT_DIR = "/content/drive/MyDrive/Thesis/Data"
FEEDBACK_DIR = join(PROJECT_DIR, "Feedback data")
PROCESSED_DIR = join(PROJECT_DIR, "Processed")

Mounted at /content/drive


Loading the dataset with all transaction data:


In [27]:
sr_full_feedback = pd.read_csv(join(FEEDBACK_DIR,"sr_full_feedback_data.csv"), encoding='latin1')
ab_full_feedback = pd.read_csv(join(FEEDBACK_DIR,"ab_full_feedback_data.csv"), encoding='latin1')
ha_full_feedback = pd.read_csv(join(FEEDBACK_DIR,"ha_full_feedback_data.csv"), encoding='latin1')

Loading the dataset with selected transaction data:

In [28]:
sr_text_feedback = pd.read_csv(join(FEEDBACK_DIR,"sr_selected_feedback_data.csv"), encoding='latin1')
ab_text_feedback = pd.read_csv(join(FEEDBACK_DIR,"ab_selected_feedback_data.csv"), encoding='latin1')
ha_text_feedback = pd.read_csv(join(FEEDBACK_DIR,"ha_selected_feedback_data.csv"), encoding='latin1')

Loading the dataset with manually coded
 data:



In [33]:
manually_coded = pd.read_csv(join(PROJECT_DIR,"Annotation data/manual_coding_majority_vote_2000_texts.csv"), encoding='latin1')
annotation_records = pd.read_csv(join(PROJECT_DIR,"Annotation data/manual_coding_6000_texts.csv"), encoding='latin1')

# 1 DESCRIPTIVE STATISTICS

## FULL FEEDBACK

**Share of feedback with textual data**

Explore whether the feedback is default text or have NaN **values**:


In [ ]:
# Define default text lists
sr_default_text = [
    "Leave feedback here.",
    "Leave feedback here",
    "Leave Feedback Here",
    "Leave Feedback here",
    "leave feedback here",
    "leave feedback here.",
    "Leave feedback here,",
    "Leave feedback here!",
    "Leave feedback"
]
ab_default_text = ["No comment"]
ha_default_text = ["---", "--", "-"]

# Count NaN values in 'comment'
sr_nan_count = sr_full_feedback['comment'].isna().sum()
ab_nan_count = ab_full_feedback['comment'].isna().sum()
ha_nan_count = ha_full_feedback['comment'].isna().sum()

print(f"""-----------------------
Number of NaN values in Silk Road comments: {sr_nan_count}
Number of NaN values in AlphaBay comments: {ab_nan_count}
Number of NaN values in Hansa comments: {ha_nan_count}
-----------------------""")

# Apply default text conditions
sr_full_feedback['default_text'] = np.where(sr_full_feedback['comment'].isin(sr_default_text), True, False)
ab_full_feedback['default_text'] = np.where(ab_full_feedback['comment'].isin(ab_default_text), True, False)
ha_full_feedback['default_text'] = np.where(ha_full_feedback['comment'].isin(ha_default_text), True, False)

sr_default_com_count = sr_full_feedback['default_text'].sum()
ab_default_com_count = ab_full_feedback['default_text'].sum()
ha_default_com_count = ha_full_feedback['default_text'].sum()

print(f"""-----------------------
Number of default comments in Silk Road: {sr_default_com_count}
Number of default comments in AlphaBay: {ab_default_com_count}
Number of default comments in Hansa: {ha_default_com_count}
-----------------------""")

# Calculate proportions of non-NaN and False default_text
sr_total_count = len(sr_full_feedback)
ab_total_count = len(ab_full_feedback)
ha_total_count = len(ha_full_feedback)

prop_valid_non_default_sr = round(((sr_full_feedback['comment'].notna()) & (sr_full_feedback['default_text'] == False)).sum() / sr_total_count * 100, 2)
prop_valid_non_default_ab = round(((ab_full_feedback['comment'].notna()) & (ab_full_feedback['default_text'] == False)).sum() / ab_total_count * 100, 2)
prop_valid_non_default_ha = round(((ha_full_feedback['comment'].notna()) & (ha_full_feedback['default_text'] == False)).sum() / ha_total_count * 100, 2)

print(f"""-----------------------
Proportion of valid non-default comments in Silk Road: {prop_valid_non_default_sr}%
Proportion of valid non-default comments in AlphaBay: {prop_valid_non_default_ab}%
Proportion of valid non-default comments in Hansa: {prop_valid_non_default_ha}%
-----------------------""")

-----------------------
Number of NaN values in Silk Road comments: 1700
Number of NaN values in AlphaBay comments: 9
Number of NaN values in Hansa comments: 0
-----------------------
-----------------------
Number of default comments in Silk Road: 3458
Number of default comments in AlphaBay: 439714
Number of default comments in Hansa: 16246
-----------------------
-----------------------
Proportion of valid non-default comments in Silk Road: 97.21%
Proportion of valid non-default comments in AlphaBay: 81.38%
Proportion of valid non-default comments in Hansa: 62.08%
-----------------------


**Share of pos/neu/neg transactions**

In [ ]:
# Recode ratings
sr_full_feedback['rating_recoded'] = np.where(sr_full_feedback['rating'] == 5, 'Positive',
                                              np.where(sr_full_feedback['rating'] == 1, 'Negative', 'Neutral'))
ha_full_feedback['rating_recoded'] = np.where(ha_full_feedback['rating'] == 'success', 'Positive',
                                              np.where(ha_full_feedback['rating'] == 'danger', 'Negative', 'Neutral'))

# Calculate proportions of ratings
prop_rating_sr = round(sr_full_feedback['rating_recoded'].value_counts(normalize=True) * 100, 2)
prop_rating_ab = round(ab_full_feedback['rating'].value_counts(normalize=True) * 100, 2)
prop_rating_ha = round(ha_full_feedback['rating_recoded'].value_counts(normalize=True) * 100, 2)

# Prepare table
colnames_table = ["SilkRoad", "AlphaBay", "Hansa"]
rownames_table = ["Positive", "Neutral", "Negative", "With text", "(N =)"]

rows1_3 = pd.concat([prop_rating_sr, prop_rating_ab, prop_rating_ha], axis=1)
rows1_3.columns = colnames_table
rows1_3 = rows1_3.loc[['Negative', 'Neutral', 'Positive'], :]

row4 = pd.DataFrame({
    "SilkRoad": [prop_valid_non_default_sr],
    "AlphaBay": [prop_valid_non_default_ab],
    "Hansa": [prop_valid_non_default_ha]
}, index=["With text"])

count_row5 = pd.DataFrame({
    "SilkRoad": [((sr_full_feedback['comment'].notna()) & (sr_full_feedback['default_text'] == False)).sum()],
    "AlphaBay": [((ab_full_feedback['comment'].notna()) & (ab_full_feedback['default_text'] == False)).sum()],
    "Hansa": [((ha_full_feedback['comment'].notna()) & (ha_full_feedback['default_text'] == False)).sum()]
}, index=["(N =)"])

table = pd.concat([rows1_3, row4,count_row5])

# Apply the percentage format to the first four rows
table.iloc[:4] = table.iloc[:4].applymap(lambda x: f"{x}%" if pd.notnull(x) else x)

# Format the fifth row with thousand separators
table.iloc[4] = table.iloc[4].apply(lambda x: f"{x:,.0f}".replace(",", " ") if pd.notnull(x) else x)

table.insert(0, 'Feedbacks', rownames_table)

table

Feedbacks SilkRoad   AlphaBay   Hansa
Negative    Positive    1.11%      2.56%   1.07%
Neutral      Neutral    2.39%      1.21%   2.42%
Positive    Negative    96.5%     96.23%  96.52%
With text  With text   97.21%     81.38%  62.08%
(N =)          (N =)  179 593  1 921 591  26 597

Creating table

In [ ]:
# Create a Word document
doc = Document()
table_doc = doc.add_table(rows=table.shape[0]+1, cols=table.shape[1])

# Add headers
hdr_cells = table_doc.rows[0].cells
for i, column_name in enumerate(table.columns):
    hdr_cells[i].text = column_name

# Add the data
for i, row in enumerate(table.itertuples(), start=1):
    row_cells = table_doc.rows[i].cells
    for j, value in enumerate(row[1:], start=0):  # start=1 to skip the Index column
        row_cells[j].text = str(value)

# Style the table
for row in table_doc.rows:
    for cell in row.cells:
        cell.text = cell.text

# Save the document
doc.save(join(PROJECT_DIR, "Results/feedback_proportions.docx"))

## SELECTED FEEDBACK

Check if we have Null values in commnents:

In [29]:
# Count NaN values in 'comment'
sr_nan_c = sr_text_feedback['comment'].isna().sum()
ab_nan_c = ab_text_feedback['comment'].isna().sum()
ha_nan_c = ha_text_feedback['comment'].isna().sum()

print(f"""-----------------------
Number of NaN values in Silk Road comments: {sr_nan_c}
Number of NaN values in AlphaBay comments: {ab_nan_c}
Number of NaN values in Hansa comments: {ha_nan_c}
-----------------------""")

-----------------------
Number of NaN values in Silk Road comments: 2
Number of NaN values in AlphaBay comments: 382
Number of NaN values in Hansa comments: 0
-----------------------


In [30]:
sr_text_feedback = sr_text_feedback.loc[sr_text_feedback['comment'].notna()]
ab_text_feedback = ab_text_feedback.loc[ab_text_feedback['comment'].notna()]

Determine observations containing symbols only:


In [ ]:
# function that checks whether a comment contains letters
def symbols_only(comment):
    if isinstance(comment, str) and re.search("[A-Za-z]+", comment):
        return "no"
    else:
        return "yes"

sr_text_feedback['symbols_only'] = sr_text_feedback['comment'].apply(symbols_only)
ab_text_feedback['symbols_only'] = ab_text_feedback['comment'].apply(symbols_only)
ha_text_feedback['symbols_only'] = ha_text_feedback['comment'].apply(symbols_only)

For SilkRoad, removing 3 feedbacks of extreme length that just repeatedly
state "SCAMMER SCAMMER":

Ana delete only indices: (65322, 65316) [minus 1 since R count from 1], but  decided to delete also this text: "SCAMMER SCAMMER SCAMMER"


In [ ]:
indices_to_remove = ['sr67783', 'sr67776', 'sr67782']
sr_text_feedback = sr_text_feedback.drop(sr_text_feedback[sr_text_feedback['global_id'].isin(indices_to_remove)].index)

Get the statistics:


In [ ]:
datasets = ["sr_text_feedback", "ab_text_feedback", "ha_text_feedback"]
abb = ["sr", "ab", "ha"]

# Function to check if the comment contains letters
def symbols_only(comment):
    if isinstance(comment, str) and re.search("[A-Za-z]+", comment):
        return "no"
    else:
        return "yes"

# Processing datasets
for i, dataset_name in enumerate(datasets):
    dataset = locals()[dataset_name]
    filtered_dataset = dataset.loc[dataset['symbols_only'] == "no"].copy()

    # Number of words in the comment
    filtered_dataset['n_words'] = filtered_dataset['comment'].apply(lambda x: len(re.findall(r'\b\w+\b', x)))

    # len(text.split())

    # Number of characters with punctuation
    filtered_dataset['char_punct'] = filtered_dataset['comment'].apply(lambda x: len(re.findall(r"[^ ]", x)))

    # Number of characters without punctuation
    filtered_dataset['char_no_punct'] = filtered_dataset['comment'].apply(lambda x: len(re.findall(r"[a-zA-Z0-9]", x)))

    # Number of punctuation characters
    filtered_dataset['punct'] = filtered_dataset['char_punct'] - filtered_dataset['char_no_punct']

    text_stats = filtered_dataset[['global_id', 'comment', 'rating', 'n_words', 'char_punct', 'char_no_punct', 'punct']]

    locals()[f"{abb[i]}_text_statistics"] = text_stats


In [ ]:
def calculate_descriptives(dataframe, columns_to_exclude):
    columns = [col for col in dataframe.columns if col not in columns_to_exclude]
    data = dataframe[columns]
    results = {
        "Mean": data.mean(),
        "SD": data.std(),
        "Median": data.median(),
        "Min": data.min(),
        "Max": data.max()
    }
    desc_df = pd.DataFrame(results).T.round(2)
    desc_df.columns = columns
    return desc_df.T

columns_to_exclude = ['char_no_punct']


sr_descriptives = calculate_descriptives(sr_text_statistics.iloc[:, 3:7], columns_to_exclude)
ab_descriptives = calculate_descriptives(ab_text_statistics.iloc[:, 3:7], columns_to_exclude)
ha_descriptives = calculate_descriptives(ha_text_statistics.iloc[:, 3:7], columns_to_exclude)

Determin number of rows in each dataset:

In [ ]:
sr_nrows = sr_text_feedback.shape[0]
ab_nrows = ab_text_feedback.shape[0]
ha_nrows = ha_text_feedback.shape[0]

print(f"""-----------------------
Number of values in Silk Road: {sr_nrows}
Number of values in AlphaBay: {ab_nrows}
Number of  values in Hansa: {ha_nrows}
-----------------------""")

-----------------------
Number of values in Silk Road: 178101
Number of values in AlphaBay: 1703887
Number of  values in Hansa: 25738
-----------------------


In [ ]:
# Define the column names for descriptives
colnames_descriptives = ["Mean", "SD", "Median", "Min", "Max"]

# Create a DataFrame for the colnames_descriptives
descriptives_header = pd.DataFrame([colnames_descriptives],
                                   columns=colnames_descriptives)

# Define the first column data
first_col = [
    "N words",
    "N char (no punct.)",
    "N punct. char",
    "AlphaBay (n = 1 703 887)",
    "N words",
    "N char (no punct.)",
    "N punct. char",
    "Hansa (n = 25 738)",
    "N words",
    "N char (no punct.)",
    "N punct. char"
]

# Combine the data frames row-wise
combined_data = pd.concat([
    sr_descriptives,
    descriptives_header,
    ab_descriptives,
    descriptives_header,
    ha_descriptives
], ignore_index=True)

# Add the first column to the combined data frame
combined_data.insert(0, "First Col", first_col)

# Define the column names for the final table
colnames_table_2 = ["SilkRoad (n = 178 101)", "Mean", "SD",
                    "Median", "Min", "Max"]

# Assign the column names to the data frame
combined_data.columns = colnames_table_2

In [ ]:
combined_data

SilkRoad (n = 178 101)   Mean     SD  Median  Min     Max
0                    N words  12.66  16.64     8.0  1.0   992.0
1         N char (no punct.)   61.4  72.79    43.0  1.0  4237.0
2              N punct. char   3.31   4.42     2.0  0.0   548.0
3   AlphaBay (n = 1 703 887)   Mean     SD  Median  Min     Max
4                    N words   9.97   9.27     7.0  1.0   533.0
5         N char (no punct.)  47.19  40.91    36.0  1.0  1129.0
6              N punct. char   2.64   3.29     2.0  0.0  1056.0
7         Hansa (n = 25 738)   Mean     SD  Median  Min     Max
8                    N words   9.65  12.68     6.0  1.0   268.0
9         N char (no punct.)  46.88  57.28    30.0  1.0  1848.0
10             N punct. char   2.53   3.04     2.0  0.0    81.0

Create the table:


In [ ]:
# Create a new Word document
doc = Document()

# Add a table to the document
table = doc.add_table(rows=1, cols=len(combined_data.columns))

# Add the header row
hdr_cells = table.rows[0].cells
headers = combined_data.columns

for i, header in enumerate(headers):
    hdr_cells[i].text = header
    paragraph = hdr_cells[i].paragraphs[0]
    paragraph.alignment = WD_PARAGRAPH_ALIGNMENT.CENTER
    run = paragraph.runs[0]
    run.font.bold = True
    run.font.name = 'Times New Roman'
    run.font.size = Pt(12)

# Add the data rows
for idx, row in combined_data.iterrows():
    cells = table.add_row().cells
    for i, header in enumerate(headers):
        value = row[header]
        cells[i].text = str(value) if pd.notna(value) else ""
        paragraph = cells[i].paragraphs[0]
        paragraph.alignment = WD_PARAGRAPH_ALIGNMENT.CENTER
        run = paragraph.runs[0]
        run.font.name = 'Times New Roman'
        run.font.size = Pt(12)

# Apply bold to specific rows
bold_rows = [4, 8]
for row_idx in bold_rows:
    if row_idx < len(table.rows):
        row = table.rows[row_idx]
        for cell in row.cells:
            for paragraph in cell.paragraphs:
                for run in paragraph.runs:
                    run.font.bold = True

# Apply border to specific rows
border_rows = [4, 8]
for row_idx in border_rows:
    if row_idx < len(table.rows):
        row = table.rows[row_idx]
        for cell in row.cells:
            tc = cell._element
            tcPr = tc.get_or_add_tcPr()
            for border in ["top", "bottom"]:
                el = OxmlElement(f'w:{border}')
                el.set(qn('w:val'), 'single')
                el.set(qn('w:sz'), '8')
                el.set(qn('w:space'), '0')
                el.set(qn('w:color'), '000000')
                tcPr.append(el)

# Add footnote
doc.add_paragraph().add_run().add_break()
footnote = doc.add_paragraph(
    "The n of transactions includes texts with punctuation only, but these are excluded from the descriptive statistics for clarity."
)
footnote.add_run("1").superscript = True

# Save the document
doc.save("text_descriptives.docx")

### Dataset for prediction

Get the dataset for the final prediction (exclude manually coded sample):

In [34]:
# Anti-join
ab_for_prediction = ab_text_feedback[~ab_text_feedback['global_id'].isin(manually_coded['global_id'])]

# Tidy up the rating terminology
sr_text_feedback['rating'] = sr_text_feedback['rating'].apply(lambda x: 'Negative' if x == 1 else ('Positive' if x == 5 else 'Neutral'))

ha_text_feedback['rating'] = ha_text_feedback['rating'].apply(lambda x: 'Negative' if x == 'danger' else ('Positive' if x == 'success' else 'Neutral'))

# # Tidy up the dates
ab_for_prediction.loc[:, 'date_left'] = pd.to_datetime(ab_for_prediction['date_left'].str.slice(stop=-6), format='%b %d, %Y', errors='coerce').dt.date

# Combine datasets
prediction_dataset = pd.concat([sr_text_feedback, ab_for_prediction, ha_text_feedback], ignore_index=True)

Save the file

In [ ]:
prediction_dataset.to_csv(join(PROCESSED_DIR, "data_for_prediction.csv"), index=False)

## TOPICS OF MANUAL CODED DATA

Let's check the size of our topics in manual dataset

In [ ]:
filter_col = [col for col in manually_coded if col.startswith('t_')]
filter_col2 = [col for col in manually_coded if col.startswith('m_')]

In [ ]:
manually_coded[filter_col2].sum()

m_help_other_buyer     297
m_avoid_harm_seler      89
m_help_seller           78
m_rew_pun_seller       299
m_reach_seller          74
m_express_emo          530
m_share_facts         1396
m_other                128
dtype: int64

In [ ]:
manually_coded[filter_col].sum()

t_communication     354
t_payment            75
t_refund             96
t_price              52
t_value             112
t_shipping         1016
t_product           889
t_feedback           60
t_vendor            343
t_generic           126
t_overall           258
t_other              66
dtype: int64

# 2 Calculating the IRR


## Analyzing the distribution of raters per text

In [ ]:
print(f'Number of unique participant label:', annotation_records['participant.label'].nunique())
print(f'Number of unique participant code:', annotation_records['participant.code'].nunique())
print(f'Number of unique player email:', annotation_records['player.email'].nunique())

Number of unique participant label: 22
Number of unique participant code: 99
Number of unique player email: 35


In [ ]:
sorted(annotation_records['participant.label'].unique())

['annotator_a1',
 'annotator_a10',
 'annotator_a2',
 'annotator_a3',
 'annotator_a4',
 'annotator_a5',
 'annotator_a6',
 'annotator_a7',
 'annotator_a8',
 'annotator_a9',
 'annotator_ana',
 'annotator_b1',
 'annotator_b10',
 'annotator_b11',
 'annotator_b2',
 'annotator_b3',
 'annotator_b4',
 'annotator_b5',
 'annotator_b6',
 'annotator_b7',
 'annotator_b9',
 'annotator_wojtek']

In [ ]:
sorted(annotation_records['player.email'].unique())
# I guess that '07554973136' and '7554973136'is the same person

['07554973136',
 '7554973136',
 'AIperson',
 'DEITY',
 'F3456',
 'Haasje',
 'Iri',
 'JensKlooster',
 'Joy',
 'Kiran_SU31',
 'Mijke',
 'Nscarlet',
 'OH.',
 'Ossie',
 'SofieL',
 'SpilkoDinkov',
 'ana',
 'annotator_a1',
 'annotator_a10',
 'annotator_a7',
 'blowfish',
 'cvdm',
 'dolphin',
 'egel',
 'gadgetnr6',
 'itsme',
 'jeansinnott',
 'jessicaserena',
 'kreaseus',
 'marijana',
 'onlineresearch1',
 'phoenix23',
 'poalaaj',
 'relaxedtree',
 'wojtek']

Let's check overlapping:

In [ ]:
unique_assignments = annotation_records[['participant.label', 'player.email']].drop_duplicates().sort_values(by = 'participant.label')
unique_assignments

participant.label  player.email
4556      annotator_a1  JensKlooster
58        annotator_a1          egel
3773      annotator_a1         itsme
4739      annotator_a1      marijana
3029      annotator_a1         DEITY
...                ...           ...
1931      annotator_b7   jeansinnott
467       annotator_b7    Kiran_SU31
891       annotator_b7          egel
453       annotator_b9           Joy
0     annotator_wojtek        wojtek

[64 rows x 2 columns]

In [ ]:
unique_assignments_email = annotation_records[['player.email', 'participant.label']].drop_duplicates().sort_values(by = 'player.email')
unique_assignments_email

player.email participant.label
5615      07554973136      annotator_a2
890        7554973136      annotator_b6
889        7554973136      annotator_b5
886        7554973136      annotator_b1
2081         AIperson      annotator_a5
...               ...               ...
515   onlineresearch1      annotator_b1
1419        phoenix23      annotator_a4
2080          poalaaj      annotator_a3
530       relaxedtree      annotator_b3
0              wojtek  annotator_wojtek

[64 rows x 2 columns]

 Let's check overlapping raters and emails for texts with exactly two unique raters:

In [ ]:
# Filter texts rated by exactly two unique raters
text_rater_count = annotation_records.groupby('player.global_id')['participant.label'].nunique()
overlapping_texts = text_rater_count[text_rater_count == 2].index

# Get records with overlapping raters
overlapping_records = annotation_records[annotation_records['player.global_id'].isin(overlapping_texts)]

# Create lists of raters and emails
overlapping_raters = overlapping_records.groupby('player.global_id')['participant.label'].apply(list)
overlapping_emails = overlapping_records.groupby('player.global_id')['player.email'].apply(list)

# Get unique text values for each 'player.global_id'
text_data = annotation_records[['player.global_id', 'player.text']].drop_duplicates().set_index('player.global_id')

# Merge the lists of raters and emails with the text data
result_df = overlapping_raters.to_frame(name='raters').join(overlapping_emails.to_frame(name='emails')).join(text_data)
result_df.reset_index(inplace=True)
result_df

player.global_id                                            raters  \
0         ao1008851       [annotator_ana, annotator_a4, annotator_a4]   
1         ao1021586       [annotator_b1, annotator_ana, annotator_b1]   
2         ao1151864       [annotator_ana, annotator_a3, annotator_a3]   
3         ao1226263       [annotator_ana, annotator_a2, annotator_a2]   
4         ao1369676       [annotator_b1, annotator_ana, annotator_b1]   
5         ao1398709       [annotator_b1, annotator_ana, annotator_b1]   
6         ao1518297       [annotator_ana, annotator_a2, annotator_a2]   
7         ao1521647       [annotator_ana, annotator_a4, annotator_a4]   
8         ao1600017       [annotator_ana, annotator_a1, annotator_a1]   
9         ao1943615       [annotator_ana, annotator_a2, annotator_a2]   
10        ao2006410       [annotator_b2, annotator_b2, annotator_ana]   
11        ao2027746       [annotator_ana, annotator_a4, annotator_a4]   
12        ao2039156       [annotator_ana, annotator_b1, annotator_b1]   
13        ao2120815        [annotator_a2, annotator_a1, annotator_a2]   
14        ao2158984       [annotator_ana, annotator_b2, annotator_b2]   
15        ao2164050       [annotator_b1, annotator_ana, annotator_b1]   
16        ao2204042       [annotator_ana, annotator_a2, annotator_a2]   
17        ao2208608       [annotator_ana, annotator_a1, annotator_a1]   
18        ao2215240       [annotator_ana, annotator_a3, annotator_a3]   
19        ao2218717       [annotator_ana, annotator_a5, annotator_a5]   
20        ao2219114       [annotator_ana, annotator_a2, annotator_a2]   
21        ao2220004       [annotator_ana, annotator_b1, annotator_b1]   
22        ao2222159       [annotator_ana, annotator_a5, annotator_a5]   
23        ao2225108       [annotator_ana, annotator_a5, annotator_a5]   
24        ao2225194       [annotator_ana, annotator_a2, annotator_a2]   
25        ao2234640       [annotator_ana, annotator_a4, annotator_a4]   
26        ao2235518       [annotator_ana, annotator_a4, annotator_a4]   
27        ao2236140       [annotator_ana, annotator_a4, annotator_a4]   
28        ao2236382       [annotator_a4, annotator_a4, annotator_ana]   
29        ao2237860       [annotator_ana, annotator_b1, annotator_b1]   
30        ao2238923  [annotator_ana, annotator_ana, annotator_wojtek]   
31        ao2247136       [annotator_ana, annotator_a4, annotator_a4]   
32        ao2248035       [annotator_ana, annotator_a3, annotator_a3]   
33         ao310295       [annotator_b1, annotator_b1, annotator_ana]   
34         ao339596       [annotator_ana, annotator_b1, annotator_b1]   
35         ao631193        [annotator_a1, annotator_a1, annotator_a2]   
36         ao975350       [annotator_ana, annotator_b1, annotator_b1]   
37           as4693       [annotator_ana, annotator_a2, annotator_a2]   
38          as55669       [annotator_ana, annotator_b1, annotator_b1]   
39           as8554       [annotator_ana, annotator_a2, annotator_a2]   

                             emails  \
0                   [ana, OH., Joy]   
1               [cvdm, ana, Haasje]   
2      [ana, poalaaj, SpilkoDinkov]   
3          [ana, Ossie, Kiran_SU31]   
4               [cvdm, ana, Haasje]   
5               [cvdm, ana, Haasje]   
6          [ana, Ossie, Kiran_SU31]   
7                   [ana, OH., Joy]   
8        [ana, DEITY, JensKlooster]   
9   [ana, SpilkoDinkov, Kiran_SU31]   
10             [F3456, SofieL, ana]   
11                  [ana, OH., Joy]   
12       [ana, Haasje, jeansinnott]   
13    [Ossie, marijana, Kiran_SU31]   
14       [ana, Haasje, jeansinnott]   
15       [Haasje, ana, jeansinnott]   
16  [ana, SpilkoDinkov, Kiran_SU31]   
17    [ana, marijana, JensKlooster]   
18     [ana, poalaaj, SpilkoDinkov]   
19            [ana, blowfish, egel]   
20  [ana, SpilkoDinkov, Kiran_SU31]   
21       [ana, Haasje, jeansinnott]   
22            [ana, blowfish, egel]   
23            [ana, AIperson, egel]   
24  [ana, SpilkoDinkov, Kiran_SU31

Verification of ratings comment per each of unique participan.email:

In [ ]:
annotation_records.groupby('player.email')['player.text'].count()

player.email
07554973136           1
7554973136          170
AIperson            144
DEITY                60
F3456                58
Haasje              334
Iri                   1
JensKlooster         47
Joy                 234
Kiran_SU31          286
Mijke                56
Nscarlet             44
OH.                  92
Ossie                30
SofieL               78
SpilkoDinkov        106
ana                1976
annotator_a1         56
annotator_a10        93
annotator_a7         86
blowfish             36
cvdm                 51
dolphin              84
egel                294
gadgetnr6            45
itsme                54
jeansinnott         255
jessicaserena       125
kreaseus             41
marijana             82
onlineresearch1     109
phoenix23            27
poalaaj              66
relaxedtree          34
wojtek              742
Name: player.text, dtype: int64

Change '07554973136' to '7554973136' becuase I think is the same person:

In [ ]:
annotation_records['player.email'] = annotation_records['player.email'].replace('07554973136', '7554973136')

## Krippendorff alpha, Fleiss Kappas and PABAKS

### Motive

In [ ]:
# Prepare the data for Krippendorff's Alpha calculation
def prepare_data_for_krippendorff(data, category):
    pivoted = data.pivot_table(index='player.global_id', columns='participant.label', values=category)
    return pivoted.values

# Function to calculate Krippendorff's Alpha
def nominal_distance(x, y):
    return 0 if x == y else 1

def ordinal_distance(x, y):
    return (x - y) ** 2

def calculate_krippendorff_alpha(data, level_of_measurement):
    """
    Calculate Krippendorff's Alpha for given data and level of measurement.

    Args:
        data (numpy array): Matrix of shape (num_cases, num_raters)
        level_of_measurement (str): 'nominal' or 'ordinal'

    Returns:
        float: Krippendorff's Alpha
    """
    if level_of_measurement == 'nominal':
        distance_function = nominal_distance
    elif level_of_measurement == 'ordinal':
        distance_function = ordinal_distance
    else:
        raise ValueError("Invalid level_of_measurement. Must be 'nominal' or 'ordinal'.")

    n, k = data.shape
    assert n > 1, "Need at least two cases"
    assert k > 1, "Need at least two raters"

    # Handle missing values
    mask = np.isnan(data)
    observed_disagreement = 0
    total_ratings = 0

    # Calculate observed disagreement
    for i in range(n):
        for j in range(k):
            if not mask[i, j]:
                for l in range(j + 1, k):
                    if not mask[i, l]:
                        observed_disagreement += distance_function(data[i, j], data[i, l])
                        total_ratings += 1

    observed_disagreement /= total_ratings

    # Calculate expected disagreement
    all_ratings = data[~mask].flatten()
    unique, counts = np.unique(all_ratings, return_counts=True)
    probabilities = counts / len(all_ratings)

    expected_disagreement = 0
    for i, p_i in enumerate(probabilities):
        for j, p_j in enumerate(probabilities):
            expected_disagreement += p_i * p_j * distance_function(unique[i], unique[j])

    alpha = 1 - observed_disagreement / expected_disagreement
    return alpha

# Prepare the data for Fleiss Kappa calculation
def prepare_data_for_fleiss_kappa(data, category):
    pivoted = data.pivot_table(index='player.global_id', columns='participant.label', values=category, aggfunc='first')

    # Remove duplicate ratings by keeping the first occurrence
    pivoted = pivoted.groupby(level=0).first()

    # Ensure consistent number of raters
    mode_raters = pivoted.notna().sum(axis=1).mode()[0]

    # Create an array to hold the counts
    num_categories = 5  # Assuming ratings are integers from 0 to 4
    ratings = []

    for row in pivoted.itertuples(index=False):
        valid_ratings = [int(x) for x in row if not np.isnan(x)]
        if len(valid_ratings) != mode_raters:
            continue  # Skip rows that do not have the consistent number of raters
        counts = np.bincount(valid_ratings, minlength=num_categories)
        ratings.append(counts)

    return np.array(ratings)

# PABAK
def calculate_pabak(data):
    n, k = data.shape
    agreement_count = 0
    total_pairs = 0

    for i in range(n):
        ratings = data[i, :]
        non_nan_ratings = ratings[~np.isnan(ratings)]
        if len(non_nan_ratings) > 1:
            unique, counts = np.unique(non_nan_ratings, return_counts=True)
            agreement_count += sum(count * (count - 1) for count in counts)
            total_pairs += len(non_nan_ratings) * (len(non_nan_ratings) - 1)

    po = agreement_count / total_pairs if total_pairs > 0 else 0
    pabak = 2 * po - 1
    return pabak

In [ ]:
# Extract unique categories
motives = [
    "player.positive_negative",
    "player.help_others",
    "player.avoid_harm",
    "player.help_unrelated",
    "player.reciprocate_seller",
    "player.reach_seller",
    "player.vent_feelings",
    "player.say_facts"
]

# Corresponding labels
labels = [
  "Text polarity",
  "(12) Help other buyers...",
  "(11) Avoid harming the seller...",
  "(10) Help the seller...",
  "(14) Reward of punish the seller...",
  "(4) Reach out to the seller...",
  "(7) Express their feelings...",
  "(6) Share facts..."
]

# Calculate Krippendorff's Alpha for each motive:
alphas = []
kappas = []
pabaks = []

for category in motives:
    prepared_data = prepare_data_for_krippendorff(annotation_records, category)
    measurement = 'ordinal' if category == "player.positive_negative" else 'nominal'
    alpha_value = calculate_krippendorff_alpha(prepared_data, measurement) # my code
    alphas.append(round(alpha_value, 2))

    fleiss_data = prepare_data_for_fleiss_kappa(annotation_records, category)
    if fleiss_data.shape[0] > 0:  # Ensure there's data to process
        kappa_value = fleiss_kappa(fleiss_data)
        kappas.append(round(kappa_value, 2))
    else:
        kappas.append(np.nan)  # Append NaN if no data is available

    pabak_value = calculate_pabak(prepared_data)
    pabaks.append(round(pabak_value, 2))

# DataFrame with the results
irr_results_df = pd.DataFrame({
    'Category': motives,
    'Krippendorff_alpha': alphas,
    'Fleiss_kappa': kappas,
    'PABAK': pabaks,
    'Labels': labels
})

# Sort the DataFrame according to the specified order
sorting_indices = [0, 5, 7, 6, 3, 2, 1, 4]
irr_results_df = irr_results_df.iloc[sorting_indices]

irr_results_df

Category  Krippendorff_alpha  Fleiss_kappa  PABAK  \
0   player.positive_negative                0.89          0.73   0.66   
5        player.reach_seller                0.48          0.48   0.90   
7           player.say_facts                0.48          0.48   0.53   
6       player.vent_feelings                0.30          0.30   0.42   
3      player.help_unrelated                0.25          0.24   0.84   
2          player.avoid_harm                0.38          0.39   0.86   
1         player.help_others                0.53          0.53   0.71   
4  player.reciprocate_seller                0.40          0.40   0.65   

                                Labels  
0                        Text polarity  
5       (4) Reach out to the seller...  
7                   (6) Share facts...  
6        (7) Express their feelings...  
3              (10) Help the seller...  
2     (11) Avoid harming the seller...  
1            (12) Help other buyers...  
4  (14) Reward of punish the seller...

### Topic

In [ ]:
topics =["player.topic_communication",
        "player.topic_payment",
        "player.topic_refund",
        "player.topic_extras",
        "player.topic_value",
        "player.topic_shipping",
        "player.topic_product",
        "player.topic_feedback",
        "player.topic_vendor",
        "player.topic_generic",
        "player.topic_overall",
        "player.topic_other"]

# Calculate Krippendorff's Alpha for each topic:
alphas_topic = []
kappas_topc = []
pabaks_topic = []

for category in topics:
    prepared_data = prepare_data_for_krippendorff(annotation_records, category)
    measurement = 'ordinal' if category == "player.positive_negative" else 'nominal'
    alpha_value = calculate_krippendorff_alpha(prepared_data, measurement) # my code
    alphas_topic.append(round(alpha_value, 2))

    fleiss_data = prepare_data_for_fleiss_kappa(annotation_records, category)
    if fleiss_data.shape[0] > 0:  # Ensure there's data to process
        kappa_value = fleiss_kappa(fleiss_data)
        kappas_topc.append(round(kappa_value, 2))
    else:
        kappas_topc.append(np.nan)  # Append NaN if no data is available

    pabak_value = calculate_pabak(prepared_data)
    pabaks_topic.append(round(pabak_value, 2))

# DataFrame with the results
irr_topics = pd.DataFrame({
    'Category': topics,
    'Krippendorff_alpha': alphas_topic,
    'Fleiss_kappa': kappas_topc,
    'PABAK': pabaks_topic
})

irr_topics

Category  Krippendorff_alpha  Fleiss_kappa  PABAK
0   player.topic_communication                0.75          0.75   0.85
1         player.topic_payment                0.64          0.64   0.94
2          player.topic_refund                0.65          0.65   0.93
3          player.topic_extras                0.64          0.64   0.96
4           player.topic_value                0.60          0.60   0.90
5        player.topic_shipping                0.82          0.82   0.82
6         player.topic_product                0.74          0.74   0.74
7        player.topic_feedback                0.59          0.58   0.95
8          player.topic_vendor                0.55          0.56   0.72
9         player.topic_generic                0.77          0.77   0.94
10        player.topic_overall                0.35          0.35   0.62
11          player.topic_other                0.06          0.05   0.63

### Verification other column as rators ID

Let's try 'player.email' column as a rator index

In [ ]:
# Prepare the data for Krippendorff's Alpha calculation
def prepare_data_for_krippendorff_2(data, category):
    pivoted = data.pivot_table(index='player.global_id', columns='player.email', values=category)
    return pivoted.values

# Function to calculate Krippendorff's Alpha
def nominal_distance(x, y):
    return 0 if x == y else 1

def ordinal_distance(x, y):
    return (x - y) ** 2

def calculate_krippendorff_alpha(data, level_of_measurement):
    """
    Calculate Krippendorff's Alpha for given data and level of measurement.

    Args:
        data (numpy array): Matrix of shape (num_cases, num_raters)
        level_of_measurement (str): 'nominal' or 'ordinal'

    Returns:
        float: Krippendorff's Alpha
    """
    if level_of_measurement == 'nominal':
        distance_function = nominal_distance
    elif level_of_measurement == 'ordinal':
        distance_function = ordinal_distance
    else:
        raise ValueError("Invalid level_of_measurement. Must be 'nominal' or 'ordinal'.")

    n, k = data.shape
    assert n > 1, "Need at least two cases"
    assert k > 1, "Need at least two raters"

    # Handle missing values
    mask = np.isnan(data)
    observed_disagreement = 0
    total_ratings = 0

    # Calculate observed disagreement
    for i in range(n):
        for j in range(k):
            if not mask[i, j]:
                for l in range(j + 1, k):
                    if not mask[i, l]:
                        observed_disagreement += distance_function(data[i, j], data[i, l])
                        total_ratings += 1

    observed_disagreement /= total_ratings

    # Calculate expected disagreement
    all_ratings = data[~mask].flatten()
    unique, counts = np.unique(all_ratings, return_counts=True)
    probabilities = counts / len(all_ratings)

    expected_disagreement = 0
    for i, p_i in enumerate(probabilities):
        for j, p_j in enumerate(probabilities):
            expected_disagreement += p_i * p_j * distance_function(unique[i], unique[j])

    alpha = 1 - observed_disagreement / expected_disagreement
    return alpha

# Prepare the data for Fleiss Kappa calculation
def prepare_data_for_fleiss_kappa_2(data, category):
    pivoted = data.pivot_table(index='player.global_id', columns='player.email', values=category, aggfunc='first')

    # Remove duplicate ratings by keeping the first occurrence
    pivoted = pivoted.groupby(level=0).first()

    # Ensure consistent number of raters
    mode_raters = pivoted.notna().sum(axis=1).mode()[0]

    # Create an array to hold the counts
    num_categories = 5  # Assuming ratings are integers from 0 to 4
    ratings = []

    for row in pivoted.itertuples(index=False):
        valid_ratings = [int(x) for x in row if not np.isnan(x)]
        if len(valid_ratings) != mode_raters:
            continue  # Skip rows that do not have the consistent number of raters
        counts = np.bincount(valid_ratings, minlength=num_categories)
        ratings.append(counts)

    return np.array(ratings)

# PABAK
def calculate_pabak(data):
    n, k = data.shape
    agreement_count = 0
    total_pairs = 0

    for i in range(n):
        ratings = data[i, :]
        non_nan_ratings = ratings[~np.isnan(ratings)]
        if len(non_nan_ratings) > 1:
            unique, counts = np.unique(non_nan_ratings, return_counts=True)
            agreement_count += sum(count * (count - 1) for count in counts)
            total_pairs += len(non_nan_ratings) * (len(non_nan_ratings) - 1)

    po = agreement_count / total_pairs if total_pairs > 0 else 0
    pabak = 2 * po - 1
    return pabak

In [ ]:
# Extract unique categories
motives = [
    "player.positive_negative",
    "player.help_others",
    "player.avoid_harm",
    "player.help_unrelated",
    "player.reciprocate_seller",
    "player.reach_seller",
    "player.vent_feelings",
    "player.say_facts"
]

# Corresponding labels
labels = [
  "Text polarity",
  "(12) Help other buyers...",
  "(11) Avoid harming the seller...",
  "(10) Help the seller...",
  "(14) Reward of punish the seller...",
  "(4) Reach out to the seller...",
  "(7) Express their feelings...",
  "(6) Share facts..."
]

# Calculate Krippendorff's Alpha for each motive:
alphas = []
kappas = []
pabaks = []

for category in motives:
    prepared_data = prepare_data_for_krippendorff_2(annotation_records, category)
    measurement = 'ordinal' if category == "player.positive_negative" else 'nominal'
    alpha_value = calculate_krippendorff_alpha(prepared_data, measurement) # my code
    alphas.append(round(alpha_value, 2))

    fleiss_data = prepare_data_for_fleiss_kappa_2(annotation_records, category)
    if fleiss_data.shape[0] > 0:  # Ensure there's data to process
        kappa_value = fleiss_kappa(fleiss_data)
        kappas.append(round(kappa_value, 2))
    else:
        kappas.append(np.nan)  # Append NaN if no data is available

    pabak_value = calculate_pabak(prepared_data)
    pabaks.append(round(pabak_value, 2))

# DataFrame with the results
irr_results_df = pd.DataFrame({
    'Category': motives,
    'Krippendorff_alpha': alphas,
    'Fleiss_kappa': kappas,
    'PABAK': pabaks,
    'Labels': labels
})

# Sort the DataFrame according to the specified order
sorting_indices = [0, 5, 7, 6, 3, 2, 1, 4]
irr_results_df = irr_results_df.iloc[sorting_indices]

irr_results_df

Category  Krippendorff_alpha  Fleiss_kappa  PABAK  \
0   player.positive_negative                0.89          0.73   0.66   
5        player.reach_seller                0.48          0.48   0.90   
7           player.say_facts                0.47          0.47   0.53   
6       player.vent_feelings                0.30          0.30   0.42   
3      player.help_unrelated                0.25          0.25   0.84   
2          player.avoid_harm                0.38          0.38   0.86   
1         player.help_others                0.53          0.53   0.71   
4  player.reciprocate_seller                0.40          0.40   0.65   

                                Labels  
0                        Text polarity  
5       (4) Reach out to the seller...  
7                   (6) Share facts...  
6        (7) Express their feelings...  
3              (10) Help the seller...  
2     (11) Avoid harming the seller...  
1            (12) Help other buyers...  
4  (14) Reward of punish the seller...

# 3 Clean & Normalize categories

## Handle missing categories

First load a dataset, which we previously saved:

In [4]:
df_predict = pd.read_csv(join(PROCESSED_DIR, "data_for_prediction.csv"), encoding='latin1')
df_items = pd.read_csv(join(FEEDBACK_DIR,"ab_items.csv"), encoding='latin1')
df_items_sr = pd.read_csv(join(FEEDBACK_DIR,"sr_items.csv"), encoding='latin1')

<ipython-input-4-d71a8bfe9604>:1: DtypeWarning: Columns (4,5) have mixed types. Specify dtype option on import or set low_memory=False.
  df_predict = pd.read_csv(join(PROCESSED_DIR, "data_for_prediction.csv"), encoding='latin1')


Let's check howe many rows don't have categories:

In [5]:
num = df_predict['category'].isna().sum()
percentage = (df_predict['category'].isna().sum() / df_predict.shape[0]) * 100
print(f"Number of missing values in category: {num} ({percentage:.2f}%)")

Number of missing values in category: 483774 (25.39%)


In [9]:
df_predict[df_predict['category'].isna()==True]['dataset'].value_counts()

dataset
ab    483774
Name: count, dtype: int64

In [10]:
df_predict[df_predict['category'].isna()==True]

seller   date_left  buyer  \
178134                 HumboldtFarms  2016-03-21  d...g   
178209                 HumboldtFarms  2016-02-11  t...e   
178223                 HumboldtFarms  2016-09-16  l...g   
178234                 HumboldtFarms  2016-06-18  i...n   
178277                 HumboldtFarms  2016-09-01  d...n   
...                              ...         ...    ...   
1879939                      Telek0m  2017-02-22  n...m   
1879942  the-original-pharmacypowder         NaN  d...s   
1879946                       TheCEO  2017-05-04  d...2   
1879972                       weedim  2016-05-15  p...9   
1879987                      ZipLine  2015-08-19  f...e   

                                                   comment    itemid  price  \
178134   5/5 everything was perfect.  Great product, gr...   40256.0   75.0   
178209   Very nice stealth, amazing product, extremely ...   67535.0   32.0   
178223    Fast shipping and great quality. Thanks Humbolt.   70404.0  800.0   
178234   took 6 days but well worth the wait, this stuf...   67546.0  242.0   
178277   Twice as long to pack and ship out, not like o...   70404.0  807.0   
...                                                    ...       ...    ...   
1879939  Original Medikamente, annehmbare Lieferzeit (3...  293365.0    3.2   
1879942  12 day no show with no response for the last c...  168444.0   10.0   
1879946  Received 4dd. Was 2.1 grams short. A lot more ...  339283.0  410.0   
1879972                         FE for my trusted supplier  150199.0    6.9   
1879987  great seller!!! program was put in the proper ...   20763.0    2.5   

           rating category  global_id dataset symbols_only  
178134   Positive      NaN   ao988317      ab           no  
178209   Positive      NaN   ao991751      ab           no  
178223   Positive      NaN   ao970315      ab           no  
178234   Positive      NaN   ao979637      ab           no  
178277   Positive      NaN   ao971902      ab           no  
...           ...      ...        ...     ...          ...  
1879939  Positive      NaN  ao1831950      ab           no  
1879942  Negative      NaN  ao2243180      ab           no  
1879946   Neutral      NaN  ao2210479      ab           no  
1879972  Positive      NaN  ao2117649      ab           no  
1879987  Positive      NaN  ao2182145      ab           no  

[483774 rows x 11 columns]

In [ ]:
df_nocategory = df_predict[df_predict['category'].isna()== True]
# convert 'itemid' to integer
df_nocategory.loc[:,'itemid'] = df_nocategory['itemid'].astype(float).astype(pd.Int64Dtype())

print(f"Number of unique items which don't have a category:", df_nocategory['itemid'].nunique() )
items_list = df_nocategory['itemid'].unique()

Number of unique items which don't have a category: 51458


Filtering items which don't have category:

In [ ]:
df_items_filter = df_items[df_items['itemid'].isin(items_list)]
df_items_filter

Empty DataFrame
Columns: [itemid, seller, name, description, class, quantityleft, endsin, shipsfrom, shipsto, payment, price, cat1, cat2, cat3, shipping_is_0, date_le, n_sold, since, vendor_level, trust_level]
Index: []

So we don't have these itemid in the Item dataset

But we see that the rest of items which have category exist in this dataset, but not of them. We miss in Items dataset around 109 items whcih are in df_predict

In [ ]:
common_items = set(df_items['itemid']).intersection(set(df_predict[df_predict['dataset']== 'ab']['itemid'].unique()))
print(f"Number of common elements: {len(common_items)}")
if len(common_items) > 0:
    print("Example of common elements:", list(common_items)[:10])
else:
    print("Missing common elements.")

Number of common elements: 59760
Example of common elements: [131077.0, 131083.0, 131088.0, 262161.0, 131089.0, 131092.0, 262166.0, 262169.0, 262170.0, 31.0]


In [ ]:
df_predict[(df_predict['category'].isna()== False) & (df_predict['dataset']== 'ab')]['itemid'].nunique()

59869

We can also take a look at the seller, maybe we can confirm that they sold only products from one category:

In [ ]:
seller_list = df_nocategory['seller'].unique()
df_items_seller = df_items[df_items['seller'].isin(seller_list)]
print(f"Number of unique sellers without category in our dataset: {seller_list.shape[0]}")
print(f"Number of sellers from our list found in the items dataset: {df_items_seller['seller'].nunique()}")

Number of unique sellers without category in our dataset: 4330
Number of sellers from our list found in the items dataset: 3836


In [ ]:
df_items_seller.groupby(['seller', 'cat1']).size().reset_index(name='count')

seller               cat1  count
0                $cotty              Fraud      1
1        --StrongWeed--  Drugs & Chemicals     13
2                -ACME-  Drugs & Chemicals     14
3       -DutchDynamite-  Drugs & Chemicals     54
4     -OrientalFinesse-  Drugs & Chemicals      4
...                 ...                ...    ...
5981               zues  Drugs & Chemicals    100
5982             zulu2u       Carded Items      1
5983             zulu2u  Drugs & Chemicals     22
5984             zulu2u              Fraud      1
5985             zulu2u           Services      1

[5986 rows x 3 columns]

Unfortunately, most of them sell products from several categories. We also cannot assume that for those sellers who have only one category in the database, the data is complete - perhaps they sell products from several categories, but we do not see it in the data.

Let's drop rows with NaN value in Category:

In [ ]:
df_predict_processed = df_predict.dropna(subset=['category'])

In [ ]:
df_predict_processed.shape

(1421953, 11)

## Normalize names

First we need to check the current category name:

In [ ]:
print(f"Number of unique values in category:" , df_predict_processed['category'].nunique())

Number of unique values in category: 226


In [ ]:
current_category = df_predict_processed['category'].unique()
current_category

Let's check what are the unique names of categories in datasets with ab_Items. Unfortunately, in SR they are defined as numbers, and in HA we only have two main ones (drugs, information) or information is missing

In [ ]:
# Combine data and select unique categories from ab_items:

df_predict_processed_filter = df_predict_processed[df_predict_processed['dataset']=='ab']
# Convert itemid column in df_predict_processed to integers
df_predict_processed_filter['itemid'] = df_predict_processed_filter['itemid'].astype(float).astype(int).astype(str)
df_items['itemid'] = df_items['itemid'].astype(str)

# Merge the entire datasets based on 'itemid'
merged_df = pd.merge(df_predict_processed_filter, df_items, on='itemid', how='left')

# Filter out records that do not have a value in 'cat1'
filtered_df = merged_df.dropna(subset=['cat1'])

# Select unique categories, keeping the first record for each category
unique_categories_df = filtered_df.drop_duplicates(subset='category', keep='first')

result_df = unique_categories_df[['category', 'cat1', 'cat2', 'cat3']]
result_df.reset_index(drop=True, inplace=True)

result_df

<ipython-input-36-ad2714538d1a>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_predict_processed_filter['itemid'] = df_predict_processed_filter['itemid'].astype(float).astype(int).astype(str)


category                cat1  \
0             Cannabis & Hashish   Drugs & Chemicals   
1                          Other    Digital Products   
2                 Legit Software    Digital Products   
3                          Dumps               Fraud   
4          Accounts & Bank Drops               Fraud   
5                  Paraphernalia   Drugs & Chemicals   
6                     Stimulants   Drugs & Chemicals   
7                        Ecstasy   Drugs & Chemicals   
8                   Psychedelics   Drugs & Chemicals   
9                  Dissociatives   Drugs & Chemicals   
10                       Opioids   Drugs & Chemicals   
11                        Benzos   Drugs & Chemicals   
12                  Prescription   Drugs & Chemicals   
13                      Fake IDs   Counterfeit Items   
14                      Steroids   Drugs & Chemicals   
15          Security & Anonymity  Guides & Tutorials   
16                   CVV & Cards               Fraud   
17  Personal Information & Scans               Fraud   
18                   Weight Loss   Drugs & Chemicals   
19                         Drugs  Guides & Tutorials   
20                       Carding            Services   
21                         Fraud  Guides & Tutorials   
22                       Hacking  Guides & Tutorials   
23                       E-Books    Digital Products   
24                           VPN  Security & Hosting   
25                         Money   Counterfeit Items   
26                         Socks  Security & Hosting   
27                       Hosting  Security & Hosting   
28             Security Software  Software & Malware   
29            Social Engineering            Services   
30                      Clothing        Carded Items   
31                   Electronics        Carded Items   
32             Botnets & Malware  Software & Malware   
33                       Digital        Carded Items   
34                Fraud Software    Digital Products   
35                  Exploit Kits  Software & Malware   
36                       Jewelry   Counterfeit Items   
37                       Tobacco   Drugs & Chemicals   
38                      Exploits  Software & Malware   
39                     Game Keys    Digital Products   
40                          Gold       Jewels & Gold   
41                  Hand Weapons             Weapons   
42                    Explosives             Weapons   
43                       Pistols             Weapons   
44                    Appliances        Carded Items   
45                    Ammunition             Weapons   
46               Long-Range Guns             Weapons   
47                        Silver       Jewels & Gold   

                            cat2                          cat3  
0             Cannabis & Hashish                       Edibles  
1                          Other                         Other  
2                 Legit Software                Legit Software  
3                          Dumps                         Dumps  
4          Accounts & Bank Drops                          Porn  
5                  Paraphernalia                 Paraphernalia  
6                     Stimulants                         Speed  
7                        Ecstasy                         Pills  
8                   Psychedelics                           LSD  
9                  Dissociatives                      Ketamine  
10                       Opioids                        Heroin  
11                        Benzos                         Pills  
12                  Prescription                  Prescription  
13                      Fake IDs                      Fake IDs  
14                      Steroids                      Steroids  
15          Security & Anonymity          Security & Anonymity  
16                   CVV & Cards                   CVV & Cards  
17  Personal Information & Scans  Personal Information & Scans  
18                   Weight Loss            

Now we can create a dictionary with categories:

In [80]:
new_categories = {
    "Drugs & Chemicals": [
        'Paraphernalia', 'Cannabis & Hashish', 'Weight Loss',  'Tobacco', 'Prescription',
        'Steroids', 'Dissociatives', 'Ecstasy', 'Psychedelics', 'Benzos', 'Stimulants',
        'Opioids', 'Weight loss',

        'Cannabis', 'Hash', '2C-P', '5-MeO-DiPT', '5-MEO-MIPT', '5-APB', '4-AcO-MIPT', 'AMT',
        '2C-C', '2C-T-7', '4-FA', '2C-T-2', '4-MEC',  '25D-NBOMe', '2C-B',  '25I-NBOMe', '2C-I',
        '2C Family', '6-APB', '3,4DMMC', '4-FMA',  '2C-T-4', '4-HO-MIPT', '4-HO-MET', '2C-D',
        '2C-E', '2-FMA', '5-MEO-AMT', '4-HO family', '25B-NBOMe', '2-FA', '4-AcO-DET', '5-MeO-DMT',
        '25C-NBOMe', 'Chemicals', 'Ephedrine', 'MDPPP',  'A-PVP', '4-AcO-DMT'
        'DOI', 'DOC', 'DOB', 'DXM', 'DOM',  'DMT', 'DOx', 'GHB', 'Barbiturates', 'Butylone',
        'Crack', 'Crystal / Powder', 'Crystal', 'Dimethocaine', 'Cocaine',  'Methylone',
        'Meth', 'Mephedrone', 'MDA', 'MXE',  'MDMA', 'MDPV', 'LSD', 'Opium', 'Ethylone',
        'Ethylphenidate', 'Steroids, PEDs', 'Shrooms',  'Speed', 'PCP', 'NBOMe', 'Adderall',
        'Ketamine','Heroin', 'Salvia', 'Weed','4-AcO-DMT', 'White'

    ],
    #      'Pills', 'Prescription',
    #     'Synthetic',   'Pharmaceuticals', 'Stimulants',

    "Fraud": [
        'Dumps', 'Accounts & Bank Drops', 'CVV & Cards', 'Personal Information & Scans',

        'Fraud Related'

        #  'Forgeries',
    ],

     "Digital Products & Services": [
         'Legit Software', 'Fraud Software', 'Game Keys', 'Other', 'E-Books',

          'Digital goods'
#         'Digital currencies',, 'CDs, DVDs, software, etc.', 'Legit Software', 'Computer equipment',
    ],

    "Security & Hosting": [
      'Hosting', 'VPN', 'Socks', 'Security & Hosting'
    ],

    "Jewels & Gold": [
        'Gold', 'Silver',

#           'Collectibles',
    ],

    "Weapons": [
        'Hand Weapons', 'Explosives', 'Pistols', 'Ammunition', 'Long-Range Guns',

        'Weaponry'

        # 'Melee', 'Lock picks'
    ],

    "Carded Items": [
        'Clothing',  'Electronics', 'Appliances', 'Digital'
    ],

    "Counterfeit Items": [
        'Fake IDs', 'Money', 'Jewelry', 'Counterfeits',  'Jewellery',

        'Apparel', 'Sunglasses', 'Handbags',  'Watches', 'Passports', 'Forgeries'
    ],

    "Guides & Tutorials": [
        'Guides & Tutorials', 'Drugs', 'Fraud', 'Hacking', 'Security & Anonymity'
    ],

    "Services": [
        'Carding', 'Social Engineering'
    ],

    "Software & Malware": [
        'Security Software',  'Botnets & Malware', 'Exploit Kits', 'Exploits'
    ]

}

#     "Books & Media": [
#         'Books', 'Physical Books', 'Magazines', 'Video', 'Newsletters',
#     ],
#     "Health & Medical": [
#         'Medical', 'Medical supplies', 'Supplements', 'Herbs & Supplements', 'Pain Relief',
#         'Pharmaceuticals', 'Antibiotics', 'Biotic materials'
#     ],
#     "Laboratory & Scientific Supplies": [
#         'Lab Supplies', 'Lab Testing', 'Cultures', 'Biotic materials'
#     ],
#     "Food & Beverages": [
#         'Food'
#     ],
#     "Fireworks & Explosives": [
#         'Fireworks', 'Explosives'
#     ],
#     "Household Items": [
#          'Hardware', 'Instruments', 'Tickets', 'Cuttings', 'Topicals'
#     ],
#     "Smoking Accessories": [
#         'Vaporizers', 'Drug paraphernalia', 'Grinders', 'Pipes', 'Scales', 'Smoking Blend', 'Glass', 'Cones', 'Packaging'
#     ],
#     "Services": [
#         'Services', 'Custom Orders', 'Invites / Accounts', 'Recipes'
#     ],
#
# }


Check if we don't miss any values:

In [81]:
reverse_dict = {}
missing_values = []
for key, values in new_categories.items():
    for value in values:
        if value in reverse_dict:
            print(f"Warning: Duplicate entry for {value}")
        reverse_dict[value] = key

# check if all values from current_category are in reverse_dict
for item in current_category:
    if item not in reverse_dict and item is not np.nan:
        missing_values.append(item)

if missing_values:
    print("Missing values in our dictionary:")
    print(missing_values)

df = pd.DataFrame(current_category, columns=['Value'])
def map_category(value):
    return reverse_dict.get(value, np.nan)

df['Category'] = df['Value'].apply(map_category)

Missing values in our dictionary:
['Pills', 'Seeds', 'Blotter', 'Digital currencies', 'Books', 'Erotica', 'Vaporizers', 'CDs, DVDs, software, etc.', 'Cubensis', 'Viagra', 'Video', 'Security', 'Accessories', 'Pain Relief', 'Edibles', 'Concentrates', 'Invites / Accounts', 'Synthetic', 'Pharmaceuticals', 'Spores', 'Services', 'Lotteries & games', 'Drug paraphernalia', 'Art', 'Computer equipment', 'Microdots', 'Supplements', 'Medical', 'Pipes', 'Metal', 'Oils', 'Newsletters', 'Glass', 'Cones', 'Grinders', 'Scales', 'TMA Family', 'San Pedro', 'Ibogain', 'Recipes', 'Atlantis', 'Medical supplies', 'SSRIs', 'Liquid', 'Fireworks', 'Custom Orders', 'Magazines', 'Tan', 'IV/IM', 'Liberty caps', 'Kink', 'Smoking Blend', 'DOI', 'Packaging', 'Mexicana', 'Papers', 'Caffeine', 'Security, defense', 'Lab Supplies', 'Loans', 'Herbs & Supplements', 'Mescaline', 'Lock picks', 'Soma', 'Tramadol', 'Melee', 'Cultures', 'Biotic materials', 'Lab Testing', 'Physical Books', 'Instruments', 'Tickets', 'Collectibles

In [82]:
df

Value            Category
0              Cannabis   Drugs & Chemicals
1        Steroids, PEDs   Drugs & Chemicals
2                 Pills                 NaN
3                 Seeds                 NaN
4               Blotter                 NaN
..                  ...                 ...
221  Security & Hosting  Security & Hosting
222       Miscellaneous                 NaN
223        Counterfeits   Counterfeit Items
224         Europe (EU)                 NaN
225           Jewellery   Counterfeit Items

[226 rows x 2 columns]

Add new categories to dataframe

In [83]:
df_predict_processed.loc[:,'category_new'] = df_predict_processed['category'].map(reverse_dict)

Let's check the summary:

In [84]:
# Replace NaN values in 'category_new' for the grouping operation
df_predict_processed.loc[:,'category_new'] = df_predict_processed['category_new'].fillna('Unknown')

# Group by 'dataset' and 'category_new', count occurrences, and calculate percentages
grouped = df_predict_processed.groupby(['dataset', 'category_new']).size().reset_index(name='count')
grouped['percentage'] = (grouped['count'] / grouped.groupby('dataset')['count'].transform('sum') * 100).round(2).astype(str) + '%'

# Format counts with thousand separators
grouped['count'] = grouped['count'].apply(lambda x: f"{x:,}".replace(",", " "))

# Transpose the table and merge counts and percentages
transposed_combined = grouped.pivot(index='category_new', columns='dataset', values='count') + " (" + grouped.pivot(index='category_new', columns='dataset', values='percentage') + ")"

transposed_combined

dataset                                      ab               ha  \
category_new                                                       
Carded Items                      2 734 (0.22%)       34 (0.13%)   
Counterfeit Items                11 432 (0.94%)       62 (0.24%)   
Digital Products & Services      53 277 (4.37%)              NaN   
Drugs & Chemicals            1 028 601 (84.44%)              NaN   
Fraud                           100 139 (8.22%)    2 022 (7.86%)   
Guides & Tutorials               11 251 (0.92%)  17 874 (69.45%)   
Jewels & Gold                       421 (0.03%)              NaN   
Security & Hosting                3 431 (0.28%)      312 (1.21%)   
Services                          3 257 (0.27%)              NaN   
Software & Malware                2 268 (0.19%)              NaN   
Unknown                                     NaN   5 434 (21.11%)   
Weapons                           1 303 (0.11%)              NaN   

dataset                                   sr  
category_new                                  
Carded Items                     126 (0.07%)  
Counterfeit Items               5 166 (2.9%)  
Digital Products & Services    3 806 (2.14%)  
Drugs & Chemicals            121 639 (68.3%)  
Fraud                                    NaN  
Guides & Tutorials             5 112 (2.87%)  
Jewels & Gold                     26 (0.01%)  
Security & Hosting                       NaN  
Services                                 NaN  
Software & Malware                       NaN  
Unknown                      42 219 (23.71%)  
Weapons                             7 (0.0%)

In [85]:
# Group by 'category_new' and count occurrences
grouped = df_predict_processed.groupby(['category_new']).size().reset_index(name='count')
grouped['percentage'] = (grouped['count'] / grouped['count'].sum() * 100).round(2).astype(str) + '%'

# Format counts with thousand separators
grouped['count'] = grouped['count'].apply(lambda x: f"{x:,}".replace(",", " "))

grouped

category_new      count percentage
0                  Carded Items      2 894       0.2%
1             Counterfeit Items     16 660      1.17%
2   Digital Products & Services     57 083      4.01%
3             Drugs & Chemicals  1 150 240     80.89%
4                         Fraud    102 161      7.18%
5            Guides & Tutorials     34 237      2.41%
6                 Jewels & Gold        447      0.03%
7            Security & Hosting      3 743      0.26%
8                      Services      3 257      0.23%
9            Software & Malware      2 268      0.16%
10                      Unknown     47 653      3.35%
11                      Weapons      1 310      0.09%

In [86]:
test = df_predict_processed[df_predict_processed['category_new']=='Unknown']
test.groupby('category').count()

seller  date_left  buyer  comment  itemid  price  rating  \
category                                                                
Accessories     126        126    126      126     126    126     126   
Alcohol           4          4      4        4       4      4       4   
Antibiotics      14         14     14       14      14     14      14   
Art              24         24     24       24      24     24      24   
Atlantis          6          6      6        6       6      6       6   
...             ...        ...    ...      ...     ...    ...     ...   
Tramadol         39         39     39       39      39     39      39   
Vaporizers        9          9      9        9       9      9       9   
Viagra          599        599    599      599     599    599     599   
Video           265        265    265      265     265    265     265   
azurescens        1          1      1        1       1      1       1   

             global_id  dataset  symbols_only  category_new  
category                                                     
Accessories        126      126           126           126  
Alcohol              4        4             4             4  
Antibiotics         14       14            14            14  
Art                 24       24            24            24  
Atlantis             6        6             6             6  
...                ...      ...           ...           ...  
Tramadol            39       39            39            39  
Vaporizers           9        9             9             9  
Viagra             599      599           599           599  
Video              265      265           265           265  
azurescens           1        1             1             1  

[89 rows x 11 columns]

In [88]:
test[test['category']=='Medical supplies']

seller   date_left buyer  \
11583   1e830ba54f  2012-03-28  none   
25520   cf6fa9d0db  2012-06-20  none   
25521   cf6fa9d0db  2012-06-27  none   
25522   cf6fa9d0db  2012-03-29  none   
25523   cf6fa9d0db  2012-02-28  none   
35920   f9af523a6e  2012-07-07  none   
35921   f9af523a6e  2012-07-12  none   
118953  11a18e4388  2012-04-18  none   
118954  11a18e4388  2012-04-16  none   
118955  11a18e4388  2012-05-04  none   
153480  cf6fa9d0db  2012-07-03  none   
153481  cf6fa9d0db  2012-05-23  none   
153482  cf6fa9d0db  2012-05-12  none   
153483  cf6fa9d0db  2012-06-14  none   
153484  cf6fa9d0db  2012-07-15  none   
153485  cf6fa9d0db  2012-07-07  none   
153486  cf6fa9d0db  2012-05-17  none   
153487  cf6fa9d0db  2012-06-11  none   
153488  cf6fa9d0db  2012-07-23  none   
153489  cf6fa9d0db  2012-05-04  none   
153490  cf6fa9d0db  2012-06-26  none   
153491  cf6fa9d0db  2012-03-14  none   
153492  cf6fa9d0db  2012-06-23  none   
153493  cf6fa9d0db  2012-06-15  none   
153494  cf6fa9d0db  2012-05-10  none   
153495  cf6fa9d0db  2012-07-02  none   
153496  cf6fa9d0db  2012-06-27  none   
153497  cf6fa9d0db  2012-03-19  none   
153498  cf6fa9d0db  2012-05-11  none   
153499  cf6fa9d0db  2012-07-14  none   
153500  cf6fa9d0db  2012-05-16  none   
155688  cf6fa9d0db  2012-03-12  none   
155689  cf6fa9d0db  2012-06-28  none   
155690  cf6fa9d0db  2012-04-24  none   
155691  cf6fa9d0db  2012-04-24  none   
155692  cf6fa9d0db  2012-04-08  none   
155693  cf6fa9d0db  2012-07-08  none   
155694  cf6fa9d0db  2012-06-19  none   
155695  cf6fa9d0db  2012-04-30  none   
155696  cf6fa9d0db  2012-07-07  none   
155697  cf6fa9d0db  2012-05-23  none   
155698  cf6fa9d0db  2012-02-27  none   
155699  cf6fa9d0db  2012-06-01  none   
155700  cf6fa9d0db  2012-04-24  none   
155701  cf6fa9d0db  2012-06-05  none   
155702  cf6fa9d0db  2012-03-13  none   
155703  cf6fa9d0db  2012-07-08  none   
155704  cf6fa9d0db  2012-07-19  none   
155705  cf6fa9d0db  2012-05-26  none   
155706  cf6fa9d0db  2012-06-05  none   
167796  f9af523a6e  2012-07-07  none   
170057  11a18e4388  2012-03-11  none   

                                                  comment      itemid price  \
11583          FE as requested. Will update if necessary.  0e7dabc7fb  1.54   
25520                                                  A+  207bcb2b72  3.02   
25521   Extremely fast delivery (one day!), great pack...  207bcb2b72  3.02   
25522   no internet took a long time for me to finaliz...  207bcb2b72  3.02   
25523                      Fast, efficient, all good here  207bcb2b72  3.02   
35920                           great seller, sent extra!  340a22c83b  0.87   
35921   Arrived quickly to the States in 3 days. Every...  340a22c83b  0.87   
118953                   Smooth transaction. Good Seller.  a548559e4e   2.6   
118954  Got here quick. So much better that the dull u...  a548559e4e   2.6   
118955                   good quality item, great seller.  a548559e4e   2.6   
153480                                           the best  d8f46126ea  3.02   
153481  Flawless.  Arrived very quickly - hatter is al...  d8f46126ea  3.02   
153482                                  hatter is the man  d8f46126ea  3.02   
153483  Very good service, came fast and was nicely pa...  d8f46126ea  3.02   
153484                                          Thanks!!!  d8f46126ea  3.02   
153485                                         Excellent!  d8f46126ea  3.02   
153486                         edit so not to self incrim  d8f46126ea  3.02   
153487  Thanks hatter, excellent to have someone selli...  d8f46126ea  3.02   
153488                                          Excellent  d8f46126ea  3.02   
153489                                            A++++++  d8f46126ea  3.02   
153490           FAST shipping. Great experience. Thanks!  d8f46126ea  3.02   
153491                       crazy-fast.  way to be, man.  d8f46126ea  3.02   
153492                                          excellent  d8f46126ea  3.0

# 4 Preparing AB dataset

In [31]:
df_predict = pd.read_csv(join(PROCESSED_DIR, "data_for_prediction.csv"), encoding='latin1', low_memory=False)
ab_items = pd.read_csv(join(FEEDBACK_DIR,"ab_items.csv"), encoding='latin1')

Merge two dataset:

In [63]:
# filter only AB dataset
df_predict_filter = df_predict[df_predict['dataset']=='ab']
# remove duplicates from database with items
ab_items_no_duplicates = ab_items.drop_duplicates(subset=['itemid'], keep='first')

# Convert 'itemid' columns to string, handling NaN values
def is_finite(value):
    try:
        return np.isfinite(float(value))
    except (ValueError, TypeError):
        return False

mask_finite = df_predict_filter['itemid'].apply(is_finite)
df_predict_filter.loc[mask_finite, 'itemid'] = (
    df_predict_filter.loc[mask_finite, 'itemid']
    .astype(float)
    .astype(int)
    .astype(str)
)

ab_items_no_duplicates.loc[:,'itemid'] = ab_items_no_duplicates['itemid'].astype(str)

# Merge the entire datasets based on 'itemid'
ab_feedback = pd.merge(df_predict_filter,
                       ab_items_no_duplicates,
                       on=['itemid','seller'],
                       how='left')

ab_feedback = ab_feedback[ab_feedback['symbols_only']=='no']
ab_feedback_for_prediciton = ab_feedback.drop(columns=['symbols_only', 'dataset','buyer', 'cat2'])
ab_feedback_for_prediciton = ab_feedback_for_prediciton.dropna(subset=['category'])

<ipython-input-63-9ecb390f01f4>:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_predict_filter.loc[mask_finite, 'itemid'] = (


In [60]:
# filter only AB dataset
df_predict_filter = df_predict[df_predict['dataset']=='ab']
# remove duplicates from database with items
# ab_items_no_duplicates = ab_items.drop_duplicates(subset=['itemid'], keep='first')

# Convert 'itemid' columns to string, handling NaN values
def is_finite(value):
    try:
        return np.isfinite(float(value))
    except (ValueError, TypeError):
        return False

mask_finite = df_predict_filter['itemid'].apply(is_finite)
df_predict_filter.loc[mask_finite, 'itemid'] = (
    df_predict_filter.loc[mask_finite, 'itemid']
    .astype(float)
    .astype(int)
    .astype(str)
)

ab_items.loc[:,'itemid'] = ab_items_no_duplicates['itemid'].astype(str)

df_predict_filter['price'] = pd.to_numeric(df_predict_filter['price'], errors='coerce')
ab_items['price'] = pd.to_numeric(ab_items['price'], errors='coerce')
# ab_items_no_duplicates.loc[:,'itemid'] = ab_items_no_duplicates['itemid'].astype(str)

# Merge the entire datasets based on 'itemid'
ab_feedback = pd.merge(df_predict_filter,
                      #  ab_items_no_duplicates,
                       ab_items,
                       on=['itemid','seller', 'price'],
                       how='left')

ab_feedback = ab_feedback[ab_feedback['symbols_only']=='no']
ab_feedback_for_prediciton = ab_feedback.drop(columns=['symbols_only', 'dataset','buyer', 'cat2'])
ab_feedback_for_prediciton = ab_feedback_for_prediciton.dropna(subset=['category'])

<ipython-input-60-8141147f3b78>:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_predict_filter.loc[mask_finite, 'itemid'] = (
<ipython-input-60-8141147f3b78>:23: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_predict_filter['price'] = pd.to_numeric(df_predict_filter['price'], errors='coerce')


In [61]:
# Define a function to compare two columns, including NaNs
def compare_columns_with_nan(df, col1, col2):
    return df.apply(lambda x: x[col1] == x[col2] or (pd.isna(x[col1]) and pd.isna(x[col2])), axis=1)

# Apply the function to the DataFrame
comparison_with_nan = compare_columns_with_nan(ab_feedback_for_prediciton, 'seller_x', 'seller_y')

comparison_with_nan.value_counts()

KeyError: 'seller_x'

In [64]:
ab_feedback_for_prediciton.head()

seller   date_left  \
0  HumboldtFarms  2016-07-15   
1  HumboldtFarms  2017-04-10   
2  HumboldtFarms  2017-01-15   
3  HumboldtFarms  2015-11-29   
4  HumboldtFarms  2016-06-08   

                                             comment  itemid price_x  \
0  Amazing product, awesome stealth; 5dd; 2 pills...   67532    72.0   
1                                   Fast Shipping!!!  162780    45.0   
2                                               FE'd   30011    31.0   
3  Great quality and a great strain. Came overwei...   50252    97.0   
4            Love these, best way to smoke in public   30009   167.0   

     rating            category  global_id  \
0  Positive  Cannabis & Hashish   ao977151   
1  Positive  Cannabis & Hashish   ao972046   
2  Positive  Cannabis & Hashish  ao1004619   
3  Positive  Cannabis & Hashish   ao996416   
4  Positive  Cannabis & Hashish   ao980512   

                                                name  \
0  10 Canna-Caps - Cannabis Oil Capsule - EXTRA S...   
1  SALE!! 3.5g Pick-a-Strain  - Grade A++++ Top S...   
2  SALE! 1x Top-shelf 500mg CO2 extracted THC cle...   
3  7g Sour Jack - Grade A++++ Top Shelf Hydroponi...   
4  5x 500mg Premium Grade CO2 extracted THC vape ...   

                                         description  ... payment price_y  \
0  ::: Canna-Caps - Cannabis Oil Capsule - Extra ...  ...  Escrow    65.0   
1  Super Sour Diesel SFV OG Blue Dream J1 Gorilla...  ...  Escrow    45.0   
2  <U+27A4> Premium Topshelf 0.5mL Vape Cartidges...  ...  Escrow    31.0   
3  ::: Sour Jack - Grade A++++ Top Shelf Hydropon...  ...  Escrow    90.0   
4  <U+2713> Premium 0.5mL Vape Cartidge <U+2713> ...  ...  Escrow   149.0   

                cat1            cat3 shipping_is_0     date_le  n_sold  \
0  Drugs & Chemicals         Edibles           0.0  2017-06-14   485.0   
1  Drugs & Chemicals  Buds & Flowers           0.0  2017-06-14   584.0   
2  Drugs & Chemicals    Concentrates           0.0  2017-06-14  5972.0   
3  Drugs & Chemicals  Buds & Flowers           0.0  2017-06-14   222.0   
4  Drugs & Chemicals    Concentrates           0.0  2017-06-14   330.0   

        since vendor_level  trust_level  
0  2015-12-02         10.0         10.0  
1  2016-06-05         10.0         10.0  
2  2015-08-22         10.0         10.0  
3  2015-10-28         10.0         10.0  
4  2015-08-22         10.0         10.0  

[5 rows x 25 columns]

In [28]:
ab_feedback_for_prediciton.head()

seller_x   date_left  \
0  HumboldtFarms  2016-07-15   
1  HumboldtFarms  2017-04-10   
2  HumboldtFarms  2017-01-15   
3  HumboldtFarms  2015-11-29   
4  HumboldtFarms  2016-06-08   

                                             comment  itemid price_x  \
0  Amazing product, awesome stealth; 5dd; 2 pills...   67532    72.0   
1                                   Fast Shipping!!!  162780    45.0   
2                                               FE'd   30011    31.0   
3  Great quality and a great strain. Came overwei...   50252    97.0   
4            Love these, best way to smoke in public   30009   167.0   

     rating            category  global_id       seller_y  \
0  Positive  Cannabis & Hashish   ao977151  HumboldtFarms   
1  Positive  Cannabis & Hashish   ao972046  HumboldtFarms   
2  Positive  Cannabis & Hashish  ao1004619  HumboldtFarms   
3  Positive  Cannabis & Hashish   ao996416  HumboldtFarms   
4  Positive  Cannabis & Hashish   ao980512  HumboldtFarms   

                                                name  ... price_y  \
0  10 Canna-Caps - Cannabis Oil Capsule - EXTRA S...  ...    65.0   
1  SALE!! 3.5g Pick-a-Strain  - Grade A++++ Top S...  ...    45.0   
2  SALE! 1x Top-shelf 500mg CO2 extracted THC cle...  ...    31.0   
3  7g Sour Jack - Grade A++++ Top Shelf Hydroponi...  ...    90.0   
4  5x 500mg Premium Grade CO2 extracted THC vape ...  ...   149.0   

                cat1                cat2            cat3 shipping_is_0  \
0  Drugs & Chemicals  Cannabis & Hashish         Edibles           0.0   
1  Drugs & Chemicals  Cannabis & Hashish  Buds & Flowers           0.0   
2  Drugs & Chemicals  Cannabis & Hashish    Concentrates           0.0   
3  Drugs & Chemicals  Cannabis & Hashish  Buds & Flowers           0.0   
4  Drugs & Chemicals  Cannabis & Hashish    Concentrates           0.0   

      date_le  n_sold       since vendor_level trust_level  
0  2017-06-14   485.0  2015-12-02         10.0        10.0  
1  2017-06-14   584.0  2016-06-05         10.0        10.0  
2  2017-06-14  5972.0  2015-08-22         10.0        10.0  
3  2017-06-14   222.0  2015-10-28         10.0        10.0  
4  2017-06-14   330.0  2015-08-22         10.0        10.0  

[5 rows x 27 columns]

Save the file

In [69]:
ab_feedback_for_prediciton.to_csv(join(PROCESSED_DIR, "ab_feedback_for_prediciton.csv"), index=False)